Limit connection pool count

spladug opened the pull request:
We've seen apps suddenly fill connection pools with tonnes of
connections causing redis to start complaining about too many open
connections.

Hypothesis: a redis response time spike occurs and during the slowness
more and more greenlets start opening new connections.

This should prevent flooding with open connections by enforcing a cap on
the number of connections in a pool.

:eyeglasses: @kjoconnor @dellis23

spladug committed:
Limit connection pool count

We've seen apps suddenly fill connection pools with tonnes of
connections causing redis to start complaining about too many open
connections.

Hypothesis: a redis response time spike occurs and during the slowness
more and more greenlets start opening new connections.

This should prevent flooding with open connections by enforcing a cap on
the number of connections in a pool.

dellis23 commented:
Do we know how many is too many?  Or whether this number will be low enough?  What about if we scale the service up?  What happens if we reach the connection limit?

dellis23 commented:
🐟  Random questions, I'm sure you've thought about them and that this will be moving us in the right direction.

spladug commented:
:haircut: i've switched to the BlockingConnectionPool, now we'll wait up to 100ms before dying if the pool is at capacity (previously it'd just die immediately).  i've also moved the pool size to config so we can tweak it accordingly.

dellis23 commented on reddit_service_activity/__init__.py:
```diff
@@ -82,11 +82,16 @@ def make_processor(app_config):  # pragma: nocover
-    redis_pool = redis.ConnectionPool.from_url(cfg.redis.url)
+    redis_pool = redis.BlockingConnectionPool.from_url(
+        cfg.redis.url,
+        max_connections=100,
```
You created the config variable, but are you using it?

spladug replied:
thanks, didn't save the file properly :/

spladug committed:
woops, didn't save file

kjoconnor commented on reddit_service_activity/__init__.py:
```diff
@@ -82,11 +82,16 @@ def make_processor(app_config):  # pragma: nocover
+    redis_pool = redis.BlockingConnectionPool.from_url(
+        cfg.redis.url,
+        max_connections=cfg.redis.max_connections,
+        timeout=0.1,
```
🔕 could config out the timeout too? might be overkill

spladug replied:
yes, i'd like it to be. i'm gonna add support for "milliseconds" to [config.Timespan](https://reddit.github.io/baseplate/baseplate/config.html#baseplate.config.Timespan) first

kjoconnor commented:
🐟

spladug commented:
68ebc98f05701b4448f66bd590c32884fe6e712a
